core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java

```diff
@@ -0,0 +1,233 @@
+    }
+
+    private class PropertiesFileWatcherRunnable implements Runnable {
+        @Override
+        public void run() {
+            try {
+                WatchKey key;
+                while ((key = watchService.take()) != null) {
+                    for (WatchEvent<?> event : key.pollEvents()) {
+                        @SuppressWarnings("unchecked")
+                        final Path watchedPath = ((Path) key.watchable())
+                                .resolve(((WatchEvent<Path>) event).context());
+                        final Path realFilePath;
+                        try {
+                            realFilePath = watchedPath.toRealPath();
+                        } catch (IOException e) {
+                            logger.warn("skipping unable to get real path for {}", watchedPath);
+                            continue;
+                        }
+                        if (event.kind().equals(ENTRY_MODIFY) || event.kind().equals(ENTRY_CREATE)) {
+                            reloadPath(realFilePath);
+                        } else if (event.kind().equals(OVERFLOW)) {
+                            logger.debug("watch event may have been lost: {}", realFilePath);
+                            reloadPath(realFilePath);
+                        } else if (event.kind().equals(ENTRY_DELETE)) {
+                            logger.warn("ignoring deleted file: {}", realFilePath);
+                        }
+                    }
+
+                    final boolean reset = key.reset();
+                    if (!reset) {
+                        logger.warn("aborting reload properties file due to unexpected error");
+                        break;
+                    }
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                logger.warn("unexpected interruption while reloading properties file: ", e);
+            } catch (ClosedWatchServiceException e) {
+                // do nothing
+            }
+        }
+
+        private void reloadPath(Path filePath) {
+            ctxRegistry.values().stream().filter(
+                    ctx -> filePath.startsWith(ctx.dirPath)).forEach(ctx -> {
+                try {
+                    ctx.reloader.run();
+                } catch (Exception e) {
+                    logger.warn("unexpected error from listener: {} ", filePath, e);
+                }
+            });
+        }
+    }
+
+    private final class RestartableFuture {
+
+        @Nullable
+        private CompletableFuture<Void> future;
+        private final ExecutorService executor;
+        private final Supplier<Runnable> runnableSupplier;
+
+        private RestartableFuture(String namePrefix, Supplier<Runnable> runnableSupplier) {
+            executor = Executors.newSingleThreadExecutor(
+                    new ThreadFactoryBuilder().setDaemon(true)
+                                              .setNameFormat(namePrefix).build());
```

Comment:
When is this executor's thread terminated? I don't see any `shutdown` or `awaitTermination` calls. Even if the `Runnable` is finished running, the executor thread will stay alive. This causes resource leak in a container environment where an app can be reloaded.

Reply:
Ah you're right -- there will be idle threads.
When I tried starting and terminating the `ExecutorService`, it seemed to take too long (5s for each start-stop iteration) so I changed to using `Thread`s which seemed must faster.
Let me know if you have any further concerns